Assets/Shopify/UIToolkit/Components/RemoteImageLoader.cs

```diff
@@ -0,0 +1,140 @@
+namespace Shopify.UIToolkit {
+    using UnityEngine;
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+
+    public interface ImageLoader {
+        string GetError();
+
+        Texture2D GetTexture();
+
+        Dictionary<string, string> GetResponseHeaders();
+
+        IEnumerator Load(string url, Dictionary<string, string> headers);
+    }
+
+    public class UnityImageLoader : ImageLoader {
+        private WWW _www;
+
+        public string GetError() {
+            return _www.error;
+        }
+
+        public Texture2D GetTexture() {
+            return _www.texture;
+        }
+
+        public Dictionary<string, string> GetResponseHeaders() {
+            return _www.responseHeaders;
+        }
+
+        public IEnumerator Load(string url, Dictionary<string, string> headers) {
+            _www = new WWW(url, null, headers);
+            yield return _www;
+        }
+    }
+
+    /// <summary>
+    /// A behaviour that fetches and caches a remote image (JPEG/PNG) from the web.!-- from the web.
+    /// </summary>
+    public class RemoteImageLoader : MonoBehaviour {
```

Comment:
If this is a behaviour, why doesn't it collaborate directly with an Image component? If its designed more as a service to just support the image loading, I would rather see it as a plain c# object instead of a behaviour. You can still create a gameobject in the constructor and use that game object to run a coroutine if thats the reason why its a MonoBehaviour in the first place too?

Reply:
Reading the docs it looks like in order to start a coroutine from a game object I'll need to add a component to it and call `StartCoroutine` from that. Seems like I can either create a dummy behaviour or use this loader as a behaviour that references an image component. What do you think of doing the latter? The cache is shared between all instances that would be created so we could have many loaders using the same cache.